### PR TITLE
[Fix #1836] Create a cop for Spaces inside string interpolation

### LINKED ISSUE #1836: Implement a cop checking for space around interpolation
Just noticed we still haven't done this. Style guide reference [here](https://github.com/bbatsov/ruby-style-guide/pull/429).

### PR DESCRIPTION
**Pull request by glasnt**
Creates a cop based on the request in #1836 for the recent addition to the ruby-style-guide that gives an absolute consensis to the issue of whether or not spaces should be included within string interpolations (https://github.com/bbatsov/ruby-style-guide/pull/429)

### CODE REVIEW

**Comment by glasnt** [comment]
Ah. Whoops. Let me fix up travis. :person_frowning: :

**Comment by glasnt** [comment]
Not sure if that coveralls failure (-0.02%) is just a rounding error; the code I added has 100% coverage.

**Comment by jonas054** [comment]
It seems like most PRs fail in coveralls. Probably a rounding error.

This new cop should be in the `Style` category IMO, like most other space related cops. We only have one space cop in `Lint`, and that's for a special reason I've found is described in #921. Here there are no discernible dangers with the extra space, it's just a question of style.

You can squash the commits that are just corrections into the first commit. The last one (fixing up RuboCop source for the new cop) can be a separate commit.

File: config/enabled.yml [hunk 1]
```diff
@@ -948,6 +948,11 @@ Lint/SpaceBeforeFirstArg:
                  in a method call without parentheses.
   Enabled: true
 
+Lint/SpaceInsideStringInterpolation:
+  Description: 'No spaces inside "#{v}".'
```
Comment (bbatsov) [inline config/enabled.yml]: Maybe "Checks for padding/surrounding spaces inside string interpolation." would be better?

**Comment by glasnt** [comment]
No worries, @jonas054. I wasn't sure which it would be, Style or Lint. I'll flip it back around, and squash.

File: lib/rubocop/cop/lint/space_inside_string_interpolation.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      # This cop checks for whitespace within string interpolations.
+      #
+      # Good:
+      #    var = "This is the #{good} example"
+      #
+      # Bad:
+      #    var = "This is the #{ bad } example"
+      #
+      class SpaceInsideStringInterpolation < Cop
+        MSG = 'Space inside string interpolation detected.'
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/space_inside_string_interpolation.rb]: Please, add a blank line after this.

File: config/enabled.yml (hunk 1 above)
Comment (glasnt) [inline config/enabled.yml]: That does sound better. I'll change it now.

File: lib/rubocop/cop/lint/space_inside_string_interpolation.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        def on_dstr(node)
+          node.children.select { |n| n.type == :begin }.each do |begin_node|
+            final_node = begin_node.children.last
+
+            interp = final_node.loc.expression
+            if range_with_surrounding_space(interp) != interp
+              add_offense(final_node, :expression, MSG)
```
Comment (jonas054) [inline lib/rubocop/cop/lint/space_inside_string_interpolation.rb]: Like many restaurants, RuboCop uses MSG by default. So you can call `add_offense` with just the first two arguments.

File: lib/rubocop/cop/lint/space_inside_string_interpolation.rb (hunk 2 above)
Comment (jonas054) [inline lib/rubocop/cop/lint/space_inside_string_interpolation.rb]: Add an empty line before `on_dstr`.

File: spec/rubocop/cop/lint/space_inside_string_interpolation_spec.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+describe RuboCop::Cop::Lint::SpaceInsideStringInterpolation do
+  subject(:cop) { described_class.new }
+
+  context 'for ill-formatted string interpolations' do
+    let(:source) do
+      ['"#{ var}"',
+       '"#{ var }"',
+       '"#{var }"',
+       '"#{   var   }"']
```
Comment (bbatsov) [inline spec/rubocop/cop/lint/space_inside_string_interpolation_spec.rb]: A test for tab would be useful as well.

File: spec/rubocop/cop/lint/space_inside_string_interpolation_spec.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    end
+  end
+
+  context 'for well-formatted string interpolations' do
+    let(:source) do
+      ['"Variable is    #{var}      "',
+       '"  Variable is  #{var}"']
+    end
+
+    it 'doesnt register an offense for excess literal spacing' do
```
Comment (bbatsov) [inline spec/rubocop/cop/lint/space_inside_string_interpolation_spec.rb]: does not

**Comment by glasnt** [comment]
I think I squashed that properly; but I'm not sure. 

Waiting on travis to tell me that I'm still -0.02% out.

**Comment by bbatsov** [comment]
I still see 16 commits in the PR.

**Comment by glasnt** [comment]
Bah, git is being silly. I am trying to rebase the branch. Standby.

**Comment by glasnt** [comment]
Everything should be squashed properly now.

**Comment by glasnt** [comment]
.. And right now coverage/coveralls is reporting OK, but that keeps fluctuating :/

**Comment by bbatsov** [comment]
:+1: Thanks!

**Pull request merged**